### DISCUSSION TITLE
Attempted to call an undefined method named "getDoctrine" of class "App\Controller\ArtController".

### DISCUSSION THREAD

**Discussion by moatez17** [symfony/symfony#44886] [Q&A] [answered]
## I'm new to symfony and i just started a simple project where I have an entity called art and an art controller, I created the database and the migration were successful using doctrine. 

this is ArtController:
```
 namespace App\Controller;

use App\Entity\Art;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\Response;
use Doctrine\Persistence\ManagerRegistry;
use Symfony\Component\Routing\Annotation\Route;

class ArtController extends AbstractController
{
    #[Route('/')]
    public function index(ManagerRegistry $doctrine): Response
    {
        $entityManager = $this->getDoctrine()->getManger();
        $art = new Art;
        $art->setTitle('one');
        $art->setDescreption('this is a descrption');

        $entityManager->persist($art);
        $entityManager->flush();

        return $this->render('art/index.html.twig', [
            'controller_name' => 'ArtController',
        ]);
    }
}
```
getDoctrine() is an Undefined method and do not exist in AbstarctController

**Comment by mbabker** [comment, marked as answer]
As of Symfony 5.4, the `getDoctrine()` method in the abstract controller is deprecated, and it's removed in Symfony 6.

You've already injected the manager registry into your controller action, so replace the `$this->getDoctrine()` call with the `$doctrine` variable.

**Reply by moatez17** [reply]
thank you

**Reply by moatez17** [reply]
that solved the problem but now i am getting : Object of type Doctrine\Bundle\DoctrineBundle\Registry is not callable 
exception

and I am using 
```
use Doctrine\Persistence\ManagerRegistry;
```

**Reply by KingDestinyThe1st** [reply]
Experiencing similar issue. I replaced the $this->getDoctrine() call with the $doctrine variable but still got error of undefined variable $doctrine

use App\Entity\Crud;
use App\Form\CrudType;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\Routing\Annotation\Route;
use Doctrine\Bundle\DoctrineBundle\Registry;

**Reply by stof** [reply]
@KingDestinyThe1st do you have the method argument defining `$doctrine` in your code ?

**Reply by stof** [reply]
@moatez17 I suspect that you replaced `$this->getDoctrine()` with `$doctrine()` instead of `$doctrine`

**Reply by KingDestinyThe1st** [reply]
@moatez17 @stof here is my error 
"ErrorException
HTTP 500 Internal Server Error
Warning: Undefined variable $doctrine"

Here is my code

```php
<?php

namespace App\Controller;

use App\Entity\Crud;
use App\Form\CrudType;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\Routing\Annotation\Route;
use Doctrine\Bundle\DoctrineBundle\Registry;



class MainController extends AbstractController
{
    #[Route('/main', name: 'main')]
    public function index(): Response
    {
        return $this->render('main/index.html.twig', [
            'controller_name' => 'MainController',
        ]);
    }

    /**
     * @Route("create", name="create")
     */

     public function create(Request $request){
         $crud = new Crud();
         $form = $this->createForm(CrudType::class, $crud);
         $form->handleRequest($request);
         if($form->isSubmitted() && $form->isValid()){
             $em = $doctrine->getManager();
             $em->persist($crud);
             $em->flush();

             $this->addFlash('notice', 'Submitted Successfully!!');
         }

         return $this->render('main/create.html.twig',[
             'form' => $form->createView()
         ]);
     }

}
```

**Reply by stof** [reply]
and your code indeed does not inject the `$doctrine` argument, as done in the question being asked. So you need to add that part (or inject it in the constructor and use a property)

**Reply by Umer-Waqar3** [reply]
use Doctrine\Persistence\ManagerRegistry as PersistenceManagerRegistry;

    public function create( Request $request, PersistenceManagerRegistry $doctrine)
{
    $crud = new Crud();
         $form = $this->createForm(CrudType::class, $crud);
         $form->handleRequest($request);
         if($form->isSubmitted() && $form->isValid()){
             $em = $doctrine->getManager();
             $em->persist($crud);
             $em->flush();

             $this->addFlash('notice', 'Submitted Successfully!!');
         }

         return $this->render('main/create.html.twig',[
             'form' => $form->createView()
         ]);
}

**Comment by Arvind1515** [comment]
after execution given this error **Object of class Doctrine\Bundle\DoctrineBundle\Registry could not be converted to string**

`<?php

namespace App\Controller;

use App\Entity\Activity;
use Doctrine\ORM\EntityManagerInterface;
use App\Form\ActivityformType;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\Routing\Annotation\Route;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpFoundation\Request;
use Doctrine\Persistence\ManagerRegistry;
use Doctrine\ORM\EntityManager; 
use Doctrine\Persistence\ManagerRegistry as PersistenceManagerRegistry;

use Doctrine\ORM\Tools\Console\ConsoleRunner;


class FormController extends AbstractController
{   
    #[Route('/form', name: 'form')]
    public function index(Request $Request, ManagerRegistry $doctrine):Response
    {  
        $em = $this->$Doctrine->getManger();
        $Activity = New Activity();

        $form = $this->createForm(ActivityformType::class , $Activity);
        $form->handleRequest($Request);
        if($form->isSubmitted() && $form->isvalid()){
             
             $em->persist($Activity);
             $em->flush();
             return $this->addFlash('Notice','form submitted sucssesfully');
        }
            
        return $this->render('form/index.html.twig', [
            'post_form' => $form->createView(),
        ]);

        
    }
}`

**Reply by oliod** [reply]
#[Route('/form', name: 'form')]
public function index(Request $Request, ManagerRegistry $doctrine):Response
{
 $em = $doctrine->getManger();
 }

**Reply by Pablo-hub44** [reply]
did you mean getManager right?
not works for me

my code 
$user_repo = $this->getDoctrine()->getRepository(User::class);

chages 
$user_repo = $doctrine->getRepository(User::class);
not works, help

**Reply by m-arslan-hafeez** [reply]
This will help you, as of Symfony 5.4, the getDoctrine() method in the abstract controller is deprecated, and it's removed in Symfony 6. You've already injected the manager registry into your controller action, so replace the $this->getDoctrine() call with the $doctrine variable.
```

// src/Controller/IndexController.php

namespace App\Controller;

use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Annotation\Route;
use Doctrine\Persistence\ManagerRegistry;

class IndexController extends AbstractController
{
    private $doctrine;

    public function __construct(ManagerRegistry $doctrine)
    {
        $this->doctrine = $doctrine;
    }

    /**
     * @Route("/", name="homepage")
     */
    public function index(): Response
    {
        // Access Doctrine using $this->doctrine
        $entityManager = $this->doctrine->getManager();

        // Your controller logic here
        return $this->render('index.html.twig');
    }
}
```

**Comment by m-arslan-hafeez** [comment]
This will help you, as of Symfony 5.4, the getDoctrine() method in the abstract controller is deprecated, and it's removed in Symfony 6. You've already injected the manager registry into your controller action, so replace the $this->getDoctrine() call with the $doctrine variable.

```

// src/Controller/IndexController.php

namespace App\Controller;

use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Annotation\Route;
use Doctrine\Persistence\ManagerRegistry;

class IndexController extends AbstractController
{
    private $doctrine;

    public function __construct(ManagerRegistry $doctrine)
    {
        $this->doctrine = $doctrine;
    }

    /**
     * @Route("/", name="homepage")
     */
    public function index(): Response
    {
        // Access Doctrine using $this->doctrine
        $entityManager = $this->doctrine->getManager();

        // Your controller logic here
        return $this->render('index.html.twig');
    }
}
```
